File: pkg/@eserstack/posts/adapters/bluesky/mappers.ts
// Copyright 2023-present Eser Ozvataf and other contributors. All rights reserved. Apache-2.0 license.

/**
 * mappers.ts — pure functions that transform raw AT Protocol / Bluesky API
 * responses into domain entities. Nothing here reads from I/O or mutates state.
 */

import type { Post } from "../../domain/entities/post.ts";
import type { User } from "../../domain/entities/user.ts";
import { toHandle } from "../../domain/values/handle.ts";
import { toPostId } from "../../domain/values/post-id.ts";
import type { BlueskyPostView, BlueskyProfileResponse } from "./types.ts";

/**
 * Map a Bluesky post view to a domain Post entity.
 * Stores the AT-URI and CID in platformRef so reply chains can reference them.
 */
export function mapToDomainPost(raw: BlueskyPostView): Post {
  const replyRoot = raw.record.reply?.root;
  const inReplyToId = replyRoot !== undefined
    ? toPostId(replyRoot.uri)
    : undefined;

  return {
    id: toPostId(raw.uri),
    text: raw.record.text,
    authorHandle: toHandle(raw.author.handle),
    createdAt: new Date(raw.indexedAt),
    platform: "bluesky",
    platformRef: { uri: raw.uri, cid: raw.cid },
    ...(inReplyToId !== undefined && { inReplyToId }),
  };
}

/** Map a Bluesky actor profile to a domain User entity. */
export function mapToDomainUser(raw: BlueskyProfileResponse): User {
  return {
    id: raw.did,
    handle: toHandle(raw.handle),
    displayName: raw.displayName ?? raw.handle,
    platform: "bluesky",
    subscriptionTier: "free",
  };
}

File: pkg/ajan/postsfx/bluesky/mappers.go
// Copyright 2023-present Eser Ozvataf and other contributors. All rights reserved. Apache-2.0 license.

package bluesky

import (
	"strings"
	"time"

	"github.com/eser/stack/pkg/ajan/postsfx"
)

// mapPost converts a Bluesky apiPostView to the domain Post type.
func mapPost(v *apiPostView) *postsfx.Post {
	post := &postsfx.Post{
		ID:           postsfx.PostID(v.URI),
		Text:         v.Record.Text,
		AuthorHandle: postsfx.Handle(v.Author.Handle),
		Platform:     postsfx.PlatformBluesky,
		PlatformRef: map[string]string{
			"uri": v.URI,
			"cid": v.CID,
		},
	}

	if v.Record.CreatedAt != "" {
		if ts, err := time.Parse(time.RFC3339, v.Record.CreatedAt); err == nil {
			post.CreatedAt = ts
		}
	} else if v.IndexedAt != "" {
		if ts, err := time.Parse(time.RFC3339, v.IndexedAt); err == nil {
			post.CreatedAt = ts
		}
	}

	return post
}

// mapUser converts a Bluesky apiProfile to the domain User type.
func mapUser(p *apiProfile) *postsfx.User {
	return &postsfx.User{
		ID:               p.DID,
		Handle:           postsfx.Handle(p.Handle),
		DisplayName:      p.DisplayName,
		Platform:         postsfx.PlatformBluesky,
		SubscriptionTier: postsfx.SubscriptionFree, // no billing API
	}
}

// rkeyFromURI extracts the rkey from an AT URI (at://did/collection/rkey).
// Returns empty string if the URI does not have the expected format.
func rkeyFromURI(uri string) string {
	parts := strings.Split(uri, "/")
	if len(parts) > 0 {
		return parts[len(parts)-1]
	}

	return ""
}

// flattenThread recursively collects all posts in a thread view into a flat slice.
func flattenThread(view apiThreadView) []*postsfx.Post {
	var posts []*postsfx.Post //nolint:prealloc // recursive tree traversal; size unknown
	posts = append(posts, mapPost(&view.Post))

	for _, reply := range view.Replies {
		posts = append(posts, flattenThread(reply)...)
	}

	return posts
}
